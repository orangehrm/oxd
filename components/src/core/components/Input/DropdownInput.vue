<template>
  <div
    class="oxd-dropdown-container"
    @keyup.esc="closeDropdown"
    @keydown.enter.prevent="onSelectEnter"
    @keydown.down.exact.prevent="onSelectDown"
    @keydown.up.exact.prevent="onSelectUp"
  >
    <div class="oxd-dropdown-input">
      <oxd-input
        :hasError="hasError"
        :placeholder="placeholder"
        :disabled="disabled"
        @update:modelValue="onSearchUpdate"
        @click="toggleDropdown"
        @blur="onBlur"
      />
      <div class="oxd-dropdown-input-icon" v-if="!isLoading">
        <oxd-icon
          v-if="selected.length != 0 && clear"
          class="--clear"
          name="x"
          @click="onDropdownClear"
        />
        <oxd-icon
          class="--arrow"
          :name="open ? 'caret-up-fill' : 'caret-down-fill'"
          @click="toggleDropdown"
        />
      </div>
      <oxd-loading-spinner
        class="oxd-dropdown-input-loader"
        v-else
        :with-container="false"
      />
    </div>
    <transition name="transition-fade-down">
      <ul role="listbox" v-if="open" class="oxd-dropdown-options">
        <li
          v-for="(option, index) in filteredOptions"
          :ref="`optElm-${index}`"
          role="option"
          :key="`${index}-options-${option.id}`"
          :class="{
            'oxd-dropdown-options-item': true,
            '--disabled': option.disabled,
            '--selected': option.selected,
            '--focused': isFocused(index),
          }"
          @mousedown="onSelectOption(option)"
        >
          {{ option.label }}
        </li>
        <li
          v-if="filteredOptions.length === 0"
          class="oxd-dropdown-options-item --disabled"
        >
          No options
        </li>
      </ul>
    </transition>
    <div class="oxd-dropdown-selected" v-if="multiple">
      <oxd-chip
        v-for="(option, index) in selected"
        :key="`${index}-selected-${option.id}`"
        :label="option.label"
        class="oxd-dropdown-selected-chip"
      >
        <oxd-icon class="--clear" name="x" @click="onRemoveOption(option)" />
      </oxd-chip>
    </div>
  </div>
</template>

<script lang="ts">
import Chip from '@orangehrm/oxd/core/components/Chip/Chip.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import Input from '@orangehrm/oxd/core/components/Input/Input.vue';
import Spinner from '@orangehrm/oxd/core/components/Loader/Spinner.vue';
import debounce from '../../../utils/debounce';
import {
  defineComponent,
  onBeforeMount,
  PropType,
  reactive,
  toRefs,
  watch,
} from 'vue';

interface Option {
  id: number;
  label: string;
  disabled: boolean;
  selected: boolean;
}

export default defineComponent({
  name: 'oxd-dropdown-input',

  props: {
    modelValue: {},
    hasError: {
      type: Boolean,
      default: false,
    },
    multiple: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    preSelect: {
      type: Boolean,
      default: false,
    },
    clear: {
      type: Boolean,
      default: true,
    },
    lazyLoad: {
      type: Boolean,
      default: false,
    },
    disabledOptions: {
      type: Array as PropType<number[]>,
      default: [],
    },
    selectedOptions: {
      type: Array as PropType<number[]>,
      default: [],
    },
    placeholderText: {
      type: String,
      default: 'Type for hints...',
    },
    options: {
      type: [Array, Function],
      default: () => {
        return [];
      },
    },
  },

  emits: [
    'update:modelValue',
    'blur',
    'dropdown:cleared',
    'dropdown:opened',
    'dropdown:closed',
  ],

  components: {
    'oxd-chip': Chip,
    'oxd-icon': Icon,
    'oxd-input': Input,
    'oxd-loading-spinner': Spinner,
  },

  setup(props, {emit}) {
    const state = reactive({
      searchTerm: '',
      open: false,
      focused: false,
      selected: [],
      localOptions: [],
      isLoading: false,
      pointer: 0,
    });

    watch(
      () => state.selected,
      function() {
        emit('update:modelValue', state.selected);
      },
    );

    // Mark preselected options
    const preselectOptions = () => {
      if (props.preSelect) {
        if (props.selectedOptions.length > 0) {
          if (props.multiple) {
            props.selectedOptions.forEach(selOption => {
              const itemIndex = state.localOptions.findIndex(
                elem => elem.id === selOption,
              );
              if (itemIndex > -1) {
                const {id, label} = state.localOptions[itemIndex];
                state.selected.push({id, label});
              }
            });
          } else {
            const itemIndex = state.localOptions.findIndex(
              elem => elem.id === props.selectedOptions[0],
            );
            if (itemIndex > -1) {
              const {id, label} = state.localOptions[itemIndex];
              state.selected.push({id, label});
            }
          }
          emit('update:modelValue', state.selected);
        }
      }
    };

    // Add options to select dropdown
    const loadOptions = async () => {
      state.isLoading = true;
      if (Array.isArray(props.options)) {
        // If options are passed as an array set options directly
        state.localOptions = state.localOptions
          .concat(props.options)
          .filter((item, index, arr) => {
            return arr.findIndex(_item => _item.id === item.id) === index;
          });
        if (!props.lazyLoad) {
          preselectOptions();
        }
        state.isLoading = false;
      } else if (typeof props.options === 'function') {
        // If options are passed as a function, do funcation and set options
        new Promise(resolve =>
          resolve((props.options as Function)(state.searchTerm)),
        )
          .then(resolved => {
            if (resolved) {
              state.localOptions = state.localOptions
                .concat(resolved as Array<Object>)
                .filter((item, index, arr) => {
                  return arr.findIndex(_item => _item.id === item.id) === index;
                });
            }
          })
          .finally(() => {
            if (!props.lazyLoad) {
              preselectOptions();
            }
            state.isLoading = false;
          });
      }
    };

    if (!props.lazyLoad) {
      onBeforeMount(loadOptions);
    } else {
      watch(() => state.searchTerm, loadOptions);
    }

    return {
      ...toRefs(state),
    };
  },

  computed: {
    filteredOptions(): Option[] {
      const filter = new RegExp(this.searchTerm, 'i');
      return this.localOptions
        .filter(option => option.label.match(filter))
        .map(option => {
          return {
            id: option.id,
            label: option.label,
            disabled:
              this.disabledOptions.findIndex(item => item == option.id) > -1,
            selected:
              this.selected.findIndex(item => item.id == option.id) > -1,
          };
        });
    },
    placeholder(): string {
      return !this.multiple && this.selected.length != 0
        ? this.selected[0].label
        : this.placeholderText.replace(/'/g, '');
    },
  },

  methods: {
    onBlur() {
      this.closeDropdown();
      this.$emit('blur');
    },
    onSearchUpdate(value: string) {
      this.search(this, value);
    },
    search: debounce((vm, value: string) => {
      vm.searchTerm = value;
    }, 500),
    toggleDropdown() {
      if (!this.disabled) {
        if (!this.open) {
          this.openDropdown();
        } else {
          this.closeDropdown();
        }
      }
    },
    onDropdownClear() {
      this.selected = [];
      this.$emit('dropdown:cleared');
    },
    onSelectOption({id, label, disabled, selected}) {
      if (disabled || selected) {
        return;
      }
      if (!this.multiple) {
        this.selected = [];
      }
      this.selected.push({id, label});
      this.searchTerm = '';
      this.closeDropdown();
    },
    onRemoveOption(item: Option) {
      const itemIndex = this.selected.findIndex(elem => elem.id === item.id);
      if (itemIndex > -1) {
        this.selected.splice(itemIndex, 1);
      }
      this.closeDropdown();
    },
    openDropdown() {
      this.open = true;
      this.$emit('dropdown:opened');
    },
    closeDropdown() {
      this.open = false;
      this.searchTerm = '';
      this.pointer = 0;
      this.$emit('dropdown:closed');
    },
    onSelectDown() {
      if (this.open) {
        if (this.filteredOptions.length > this.pointer) {
          this.pointer++;
        } else {
          this.pointer = this.filteredOptions.length;
        }
      }
    },
    onSelectUp() {
      if (this.open) {
        if (this.pointer > 1) {
          this.pointer--;
        } else {
          this.pointer = 1;
        }
      }
    },
    onSelectEnter() {
      if (!this.open) {
        this.openDropdown();
      } else {
        if (this.pointer > 0) {
          this.onSelectOption(this.filteredOptions[this.pointer - 1]);
        }
      }
    },
    isFocused(index: number) {
      if (index === this.pointer - 1) {
        this.scrollToView(this.$refs[`optElm-${index}`]);
        return true;
      } else {
        return false;
      }
    },
    scrollToView(elm: HTMLElement) {
      elm.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'start',
      });
    },
  },
});
</script>

<style src="./dropdown-input.scss" lang="scss" scoped></style>
<style lang="scss">
.oxd-loading-spinner {
  width: 1rem !important;
  height: 1rem !important;
}
</style>
