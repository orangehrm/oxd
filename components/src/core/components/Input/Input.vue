<template>
<div :class="outerClasses" class="input-outer-wrapper">
  <div v-if="imageIcon" class="input-text-field-icon"> 
    <img :src="imageIcon" />
  </div>
  <oxd-divider v-if="imageIcon" :orientation="'vertical'" :style="{ height:'100%' }"/>
  <input
    :class="classes"
    :style="style"
    :value="modelValue"
    @focus="onFocus"
    @blur="onBlur"
    @input="onInput"
    v-bind="$attrs"
  />
</div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import Divider from '@orangehrm/oxd/core/components/Divider/Divider.vue';

export interface State {
  focused: boolean;
}

export default defineComponent({
  name: 'oxd-input',

  components :{
    'oxd-divider' : Divider
  },

  props: {
    // https://v3.vuejs.org/guide/migration/v-model.html#overview
    modelValue: {},
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    imageIcon: {
      type: String
    }
  },

  data(): State {
    return {
      focused: false,
    };
  },

  computed: {
    classes(): object {
      const isIconExists=this.imageIcon ? '--with-icon' : '';
      return {
        [`oxd-input${isIconExists}`]: true,
        [`oxd-input${isIconExists}--active`]: !this.focused,
        [`oxd-input${isIconExists}--focus`] : this.focused,
        [`oxd-input${isIconExists}--error`] : this.hasError
      };
    },
    outerClasses(): object {
      return this.imageIcon ? {
        'input-outer' : true,
        'input-outer--active': !this.focused,
        'input-outer--focus': this.focused,
        'input-outer--error': this.hasError,
      } : {'input-outer' : false};
    },
  },

  methods: {
    onFocus() {
      this.focused = true;
    },
    onBlur() {
      this.focused = false;
    },
    onInput(e: Event) {
      e.preventDefault();
      this.$emit('update:modelValue', (e.target as HTMLInputElement).value);
      this.$emit('input', e);
    },
  },
});
</script>

<style src="./input.scss" lang="scss" scoped></style>
