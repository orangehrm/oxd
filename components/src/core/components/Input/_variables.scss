// This file is part of OrangeHRM Inc
//
// Copyright (C) 2020 onwards OrangeHRM Inc
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see  http://www.gnu.org/licenses

$oxd-input-min-height: 45px !default;
$oxd-input-border--active: $oxd-input-control-border--active !default;
$oxd-input-border--focus: $oxd-input-control-border--focus !default;
$oxd-input-border--error: $oxd-input-control-border--error !default;
$oxd-input-box-shadow--focus: $oxd-input-control-box-shadow--focus !default;
$oxd-input-box-shadow--error: $oxd-input-control-box-shadow--error !default;
$oxd-input-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;

$oxd-file-input-horizontal-margin: 0.25rem !default;
$oxd-file-input-background-color: $oxd-white-color !default;
$oxd-file-button-background-color: $oxd-background-light-gray-color !default;
$oxd-file-button-color: $oxd-interface-gray-darken-1-color !default;
$oxd-file-button-font-size: 0.875rem;
$oxd-file-button-font-weight: 600;
$oxd-file-button-padding-horizontal: 1.5rem;
$oxd-file-button-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-file-button-background-color--disabled: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-file-button-height: 35px !default;
$oxd-file-button-width: 35px !default;

$oxd-checkbox-label-color: $oxd-input-control-font-color !default;
$oxd-checkbox-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-checkbox-check-font-size: 1rem !default;
$oxd-checkbox-height: 1rem !default;
$oxd-checkbox-width: 1rem !default;
$oxd-checkbox-border: 1px solid $oxd-interface-gray-lighten-1-color !default;
$oxd-checkbox-border-radius: 0.35rem !default;
$oxd-checkbox-label-margin: 0.5em !default;
$oxd-checkbox-checked-background-color: $oxd-primary-one-color !default;
$oxd-checkbox-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-switch-label-color: $oxd-input-control-font-color !default;
$oxd-switch-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-switch-width: 2rem !default;
$oxd-switch-height: calc(#{$oxd-switch-width} / 2);
$oxd-switch-radius: calc(#{$oxd-switch-height} / 2);
$oxd-switch-selector-height: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-width: calc(#{$oxd-switch-height} - 2px);
$oxd-switch-selector-radius: calc(#{$oxd-switch-selector-height} / 2);
$oxd-switch-selector-color: $oxd-white-color !default;
$oxd-switch-label-margin: 0.5em;
$oxd-switch-off-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-on-background-color: $oxd-primary-one-color !default;
$oxd-switch-on-disabled-background-color: rgba(
  $oxd-primary-one-color,
  0.5
) !default;
$oxd-switch-off-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;
$oxd-switch-disabled-background-color: $oxd-interface-gray-lighten-2-color !default;

$oxd-radio-label-color: $oxd-input-control-font-color !default;
$oxd-radio-label-color--disabled: $oxd-interface-gray-color !default;
$oxd-radio-height: 1rem !default;
$oxd-radio-width: 1rem !default;
$oxd-radio-label-margin: 0.5em !default;
$oxd-radio-checked-color: $oxd-primary-one-color !default;
$oxd-radio-checked-disabled-color: lighten(
  $oxd-radio-checked-color,
  30%
) !default;

$oxd-dropdown-text-padding: 0 0.25rem 0 0.5rem !default;
$oxd-dropdown-animation-time: 300ms !default;
$oxd-dropdown-max-height: 11.5rem !default;
$oxd-dropdown-loader-height: 1rem !default;
$oxd-dropdown-dropdown-z-index: $oxd-dropdown-z-index !default;
$oxd-dropdown-dropdown-box-shadow: 0px 5px 15px -8px rgba(0, 0, 0, 0.15) !default;
$oxd-dropdown-dropdown-padding: 4px !default;
$oxd-dropdown-dropdown-border: 1px solid $oxd-interface-gray-lighten-2-color !default;
$oxd-dropdown-dropdown-border-radius: 0.65rem !default;
$oxd-dropdown-dropdown-background: $oxd-white-color !default;
$oxd-dropdown-dropdown-background--disabled: darken(
  $oxd-background-white-shadow-color,
  3%
) !default;
$oxd-dropdown-dropdown-margin: 5px !default;
$oxd-dropdown-dropdown-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-dropdown-slot-color: $oxd-interface-gray-color !default;
$oxd-dropdown-dropdown-max-height: 224px !default;

$oxd-dropdown-option-font-size: $oxd-input-control-font-size !default;
$oxd-dropdown-option-font-family: $oxd-input-control-font-family !default;
$oxd-dropdown-option-font-weight: $oxd-input-control-font-weight !default;
$oxd-dropdown-option-padding: 4px !default;
$oxd-dropdown-option-min-height: 27px !default;
$oxd-dropdown-option-line-height: normal !default;
$oxd-dropdown-option-border-bottom: 1px solid #fbfbfc !default;
$oxd-dropdown-option-border-radius: 0.65rem !default;
$oxd-dropdown-option-color: $oxd-interface-gray-darken-1-color !default;
$oxd-dropdown-option-color--disabled: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-color--selected: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-option-background: transparent !default;
$oxd-dropdown-option-background--focused: #f6f7f9 !default;

$oxd-dropdown-chip-margin: 5px !default;
$oxd-dropdown-chip-background-color: #f6f7f9 !default;
$oxd-dropdown-chip-clear-background-color: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-chip-clear-font-color: $oxd-white-color !default;
$oxd-dropdown-chip-border-radius: 50% !default;
$oxd-dropdown-chip-area-padding: 0.5rem 0 !default;

$oxd-dropdown-arrow-background-color: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--disabled: $oxd-background-light-gray-color !default;
$oxd-dropdown-arrow-background-color--readonly: darken(
  $oxd-white-color,
  1%
) !default;
$oxd-dropdown-arrow-color--readonly: $oxd-interface-gray-lighten-1-color !default;
$oxd-dropdown-arrow-width: 35px !default;
$oxd-dropdown-arrow-height: 35px !default;
$oxd-dropdown-arrow-border-radius: 0.65rem !default;

$oxd-time-picker-period-label-width: 60px !default;
$oxd-time-picker-period-label-padding: 0.25rem 0rem !default;
$oxd-time-picker-period-horizontal-padding: 16px !default;
$oxd-time-picker-period-vertical-padding: 8px !default;
$oxd-time-picker-period-font-size: 14px !default;
$oxd-time-picker-period-font-weight: 600 !default;
$oxd-time-picker-period-font-family: $oxd-font-family !default;
$oxd-time-picker-period-font-color: $oxd-white-color !default;
$oxd-time-picker-period-border-radius: 0.5rem !default;
$oxd-time-picker-dropdown-margin: 1.5rem 1rem !default;
$oxd-time-picker-dropdown-width: 250px !default;
$oxd-time-picker-dropdown-input-width: 60px !default;
$oxd-time-picker-dropdown-seperator-padding: 0.5rem !default;

$oxd-date-input-links-margin: 0.5rem !default;
$oxd-date-input-link-padding: 5px 10px !default;
$oxd-date-input-link-border-radius: 0.75rem !default;
$oxd-date-input-link-font-family: $oxd-input-control-font-family !default;
$oxd-date-input-link-font-color: $oxd-input-control-font-color !default;
$oxd-date-input-link-font-weight: 500 !default;
$oxd-date-input-link-font-size: 14px !default;
$oxd-date-input-link-highlight-color: $oxd-primary-one-color !default;
$oxd-date-input-link-background: $oxd-background-light-gray-color !default;
