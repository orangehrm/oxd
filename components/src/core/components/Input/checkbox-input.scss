@import '../../../styles';
@import 'variables';

.oxd-checkbox-wrapper {
  & label {
    cursor: pointer;
    display: flex;
    align-content: center;
    font-family: $oxd-input-control-font-family;
    color: $oxd-checkbox-label-color;
    font-size: $oxd-input-control-font-size;
    font-weight: $oxd-input-control-font-weight;
    width: fit-content;

    &:after {
      content: '';
      display: table;
      clear: both;
    }

    & input[type='checkbox'] {
      opacity: 0;
      position: absolute;
      + .oxd-checkbox-input {
        background-color: $oxd-white-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }

      &:checked + .oxd-checkbox-input {
        background-color: $oxd-checkbox-checked-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:checked:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 1;
        }
      }

      &:disabled + .oxd-checkbox-input {
        background-color: $oxd-checkbox-disabled-background-color;
        .oxd-checkbox-input-icon {
          opacity: 0;
        }
      }
    }

    &.--disabled {
      color: $oxd-checkbox-label-color--disabled;
      cursor: unset;
    }
  }
}

.oxd-checkbox-input {
  position: relative;
  display: flex;
  border-radius: $oxd-checkbox-border-radius;
  width: $oxd-checkbox-width;
  height: $oxd-checkbox-height;

  &.--label-left {
    margin-left: $oxd-checkbox-label-margin;
  }

  &.--label-right {
    margin-right: $oxd-checkbox-label-margin;
  }

  &--active {
    border: $oxd-checkbox-border;
  }

  &--focus {
    border: $oxd-input-border--focus;
    box-shadow: $oxd-input-box-shadow--focus;
    outline: 0;
  }

  &--error {
    border: $oxd-input-border--error;
    box-shadow: $oxd-input-box-shadow--error;
  }

  &-icon {
    color: $oxd-white-color;
    font-size: $oxd-checkbox-check-font-size;
    position: absolute;
    line-height: 1.5;
    align-self: center;
  }
}
