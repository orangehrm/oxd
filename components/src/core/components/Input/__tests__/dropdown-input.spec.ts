/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */

import {mount, shallowMount} from '@vue/test-utils';
import Input from '@ohrm/oxd/core/components/Input/Input.vue';
import DropdownInput from '@ohrm/oxd/core/components/Input/DropdownInput.vue';
import Spinner from '@ohrm/oxd/core/components/Loader/Spinner.vue';

const dropdownOptions = [
  {
    id: 1,
    label: 'HR Admin',
  },
  {
    id: 2,
    label: 'ESS User',
  },
  {
    id: 3,
    label: 'Supervisor',
  },
];

const syncFunction = function () {
  return dropdownOptions;
};

const asyncFunction = async function () {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(dropdownOptions);
    }, 500);
  });
};

describe('DropdownInput.vue', () => {
  it('renders OXD Dropdown Input', () => {
    const wrapper = shallowMount(DropdownInput);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not open the dropdown when the el is clicked but the component is disabled', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        disabled: true,
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:opened')).toBeFalsy();
    expect(wrapper.vm.open).toEqual(false);
    expect(wrapper.find('.oxd-dropdown-options').exists()).toBeFalsy();
  });

  it('should open the dropdown when the el is clicked', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:opened')).toBeTruthy();
    expect(wrapper.vm.open).toEqual(true);
    expect(wrapper.find('.oxd-dropdown-options').exists()).toBeTruthy();
  });

  it('should not open the dropdown when the arrow is clicked if disabled', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        disabled: true,
      },
    });
    await wrapper.vm.$nextTick();
    wrapper.find('.oxd-icon.--arrow').trigger('click');
    expect(wrapper.find('.oxd-icon.--arrow').classes()).toContain(
      'bi-caret-down-fill',
    );
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-icon.--arrow').classes()).toContain(
      'bi-caret-down-fill',
    );
    expect(wrapper.emitted('dropdown:opened')).toBeFalsy();
    expect(wrapper.vm.open).toEqual(false);
    expect(wrapper.find('.oxd-dropdown-options').exists()).toBeFalsy();
  });

  it('should open the dropdown when the arrow is clicked if not disbled', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    await wrapper.vm.$nextTick();
    wrapper.find('.oxd-icon.--arrow').trigger('click');
    expect(wrapper.find('.oxd-icon.--arrow').classes()).toContain(
      'bi-caret-down-fill',
    );
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-icon.--arrow').classes()).toContain(
      'bi-caret-up-fill',
    );
    expect(wrapper.emitted('dropdown:opened')).toBeTruthy();
    expect(wrapper.vm.open).toEqual(true);
    expect(wrapper.find('.oxd-dropdown-options').exists()).toBeTruthy();
  });

  it('will close the dropdown and emit blur event onBlur', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:opened')).toBeTruthy();
    expect(wrapper.vm.open).toEqual(true);
    wrapper.findComponent(Input).trigger('blur');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('dropdown:closed')).toBeTruthy();
    expect(wrapper.vm.open).toEqual(false);
  });

  it('will load options from array', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    expect(wrapper.vm.localOptions.length).toEqual(3);
    expect(wrapper.vm.filteredOptions.length).toEqual(3);
  });

  it('will filter options from array', () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    wrapper.findComponent(Input).setValue('HR Admin');
    // offset for debounce
    setTimeout(() => {
      expect(wrapper.vm.localOptions.length).toEqual(3);
      expect(wrapper.vm.filteredOptions.length).toEqual(1);
    }, 600);
  });

  it('will load options from sync function', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        createOptions: syncFunction,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.localOptions.length).toEqual(3);
    expect(wrapper.vm.filteredOptions.length).toEqual(3);
  });

  it('will load options from async function', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        createOptions: asyncFunction,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isLoading).toEqual(true);
    expect(wrapper.findComponent(Spinner).exists()).toBeTruthy();
    expect(wrapper.vm.localOptions.length).toEqual(0);
    // offset for loading
    setTimeout(() => {
      expect(wrapper.vm.isLoading).toEqual(false);
      expect(wrapper.findComponent(Spinner).exists()).toBeFalsy();
      expect(wrapper.vm.localOptions.length).toEqual(3);
    }, 600);
  });

  it('will select options from dropdown', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.oxd-dropdown-options-item').length).toEqual(3);
    wrapper.findAll('.oxd-dropdown-options-item')[0].trigger('mousedown');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
  });

  it('will not select options from dropdown if disabled', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        disabledOptions: [1],
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.findAll('.oxd-dropdown-options-item').length).toEqual(3);
    wrapper.findAll('.oxd-dropdown-options-item')[0].trigger('mousedown');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.selectedOptions.length).toEqual(0);
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });

  it('will select multiple options from dropdown', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        multiple: true,
      },
    });
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    wrapper.findAll('.oxd-dropdown-options-item')[0].trigger('mousedown');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    wrapper.findComponent(Input).trigger('click');
    await wrapper.vm.$nextTick();
    wrapper.findAll('.oxd-dropdown-options-item')[1].trigger('mousedown');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
  });

  it('will clear select options from dropdown', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();
    wrapper.find('.oxd-icon.--clear').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toBeTruthy();
    expect(wrapper.emitted('dropdown:cleared')).toBeTruthy();
  });

  it('will display clear icon', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-icon.--clear').exists()).toBeTruthy();
  });

  it('will not display clear icon if not selected', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-icon.--clear').exists()).toBeFalsy();
  });

  it('will not display clear icon if clear option disabled', async () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        clear: false,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
        ],
      },
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.oxd-icon.--clear').exists()).toBeFalsy();
  });

  it('will auto preselect given options', () => {
    const wrapper = mount(DropdownInput, {
      props: {
        options: dropdownOptions,
        multiple: true,
        modelValue: [
          {
            id: 1,
            label: 'HR Admin',
          },
          {
            id: 2,
            label: 'ESS User',
          },
        ],
      },
    });
    expect(wrapper.vm.selectedOptions.length).toEqual(2);
  });
});
