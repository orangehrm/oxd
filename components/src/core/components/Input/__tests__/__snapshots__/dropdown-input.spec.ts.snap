// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownInput.vue renders OXD Dropdown Input 1`] = `
<div class="oxd-dropdown-container">
  <div class="oxd-dropdown-input">
    <oxd-input-stub haserror="false" placeholder="Type for hints..." disabled="false"></oxd-input-stub>
    <div class="oxd-dropdown-input-icon">
      <!--v-if-->
      <oxd-icon-stub name="caret-down-fill" class="--arrow"></oxd-icon-stub>
    </div>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
  <!--v-if-->
</div>
`;
